Configuration guide (#9478)
keycloak/keycloak: docs/guides/src/main/server/configuration_overview.adoc

.Command-line argument
<@kc.all parameters="--db-url-host=mykeycloakdb"/>

.Environment variable
```
export KC_DB_URL_HOST=mykeycloakdb
```

.`conf/keycloak.conf`
```
db-url-host=mykeycloakdb
```

The configuration source and the corresponding format you should use is use-case specific. That decision depends on the platform the server is deployed to
as well as on the optimizations you can get for an optimal runtime.

== Configuration Stages

The server options are narrowed to a specific command or configuration stage. This clear separation is one of the key aspects
of the server configuration and is related to a series of optimizations that are done in order to deliver the best runtime when starting and running the server.

Configuration is basically done in two stages:

* Before starting the server in order to build an optimized server image for an optimal runtime
* When starting the server

The first stage involves running the `build` command using any **build option** available from this command. On the other hand, the second stage involves
starting the server using any **configuration option** available from the `start` command.

== Configuring the server

> The clear separation between build options and configuration options is one of the key aspects

Reviewer: Is this sentence repeated on purpose? It also begins the "Configuration Stages" section.

Author: It was, but we can remove it.

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/kc.adoc" as kc>
<#import "/templates/options.adoc" as opts>
<#import "/templates/links.adoc" as links>

<@tmpl.guide
title="Configuring Keycloak"
summary="An overview of the server configuration">

In this guide, you will learn the core concepts around the server configuration and how to configure the server using the different
configuration options available. You will also learn how to properly configure the server to achieve an optimal runtime for faster
startup and low memory footprint.

== Understanding Server Options

> All the available configuration options can be found by either running the `help` command for individual commands:

Reviewer: Eliminate "the available" which adds no meaning.
Eliminate "either" because of the suggested change on the next line.

<@kc.all parameters="--help"/>

> Or by looking at all the configuration options at <@links.server id="all-config"/>.

Reviewer: "Or" is a conjuction in a sentence. 
Change to "You can instead look at all configuration options at <@links.server id="all-config"/>."

Configuration options are loaded from different sources in a specific order and they use different formats:

|===
|*Source* | *Format*

|CLI
|--<category>-<sub-category>-<property>=<value>

|Environment Variable
|KC_CATEGORY_SUB_CATEGORY_PROPERTY

|`conf/keycloak.conf`
|<category>-<sub-category>-<property>=<value>
|===

> By taking the `db-url-host` option as an example, you would set this property as follows:

Reviewer: Change "By taking" to "Given"

.Command-line argument
<@kc.all parameters="--db-url-host=mykeycloakdb"/>

.Environment variable
```
export KC_DB_URL_HOST=mykeycloakdb
```

.`conf/keycloak.conf`
```
db-url-host=mykeycloakdb
```

> The configuration source and the corresponding format you should use is use-case specific. It depends on the platform the server is deployed to

Reviewer: "It" is ambiguous. To clarify change it to "That decision...."

* When starting the server

The first stage involves running the `build` command using any **build option** available from this command. On the other hand, the second stage involves
starting the server using any **configuration option** available from the `start` command.

== Configuring the server

The clear separation between build options and configuration options is one of the key aspects
of the server configuration and related to a series of optimizations that are done in order to deliver the best runtime when starting and running the server.

The first step when configuring the server is to use the `build` command to set any build option you want to change. For instance, to choose a specific database vendor.
By looking at the individual configuration options at <@links.server id="all-config"/>, you notice
that some options are marked with an icon to indicate they are a build option. In other words, they can only be set and only take effect when running the `build` command.
You can also check all the options that require a build by looking at the `help` message of the `build` command:

<@kc.build parameters="--help"/>

> The `build` command is responsible for producing an immutable and optimized server image (similar concept to building a container image). In addition to persisting

Reviewer 2: isn't it ", which is _a_ similar ..."? ;)

> any build option you have set, this command also performs a series of optimizations to deliver the best runtime when starting and running the server. Basically,

Reviewer: Change "Basically" to "As a result"

a lot of processing that would usually happen when starting and running the server is no longer necessary and the server can start and run faster.

Some optimizations performed by the `build` command are:

> * Closed-world assumption about installed providers, hence no need to re-create the registry every time the server starts

Reviewer: change "hence" to 
which means that

* Configuration files are pre-parsed to reduce IO when starting the server
* Database specific resources are configured and prepared to run against a specific database vendor
* By persisting build options into the server image, the server does not perform any additional step to interpret configuration options and (re)configure itself

> Once you run the `build` command, you won't need to set the same build options again (they are now persisted) so that you can just `start` the server:

Reviewer: Change
again (they are now persisted) so that you can just `start` the server:
TO
again, because they are now persisted.  You can just `start` the server:

.Set any build option
<@kc.build parameters="--db=postgres"/>

.Set any configuration option to `conf/keycloak.conf`
```
db-url-host=keycloak-postgres
db-username=keycloak
db-password=change_me
hostname mykeycloak.acme.com
```

.Start the server
<@kc.start/>

== Configuring the server from a properties file

By default, the server is going to always fetch any configuration option you set in the `conf/keycloak.conf` file. When you are using a fresh distribution,
this file holds only the recommended settings for running in production which are initially commented out.

You can also specify a different configuration file by using the `[-cf|--config-file] option as follows:

<@kc.all rootParameters="-cf myconfig.conf"/>

[NOTE]
Changes to any configuration option defined in the `keycloak.conf` file that is targeted for the `build` command is ignored
if the value differs from the value used to previously run the `build` command. In this case, make sure you run the `build` command again so that
these configurations options take effect.

== Using placeholders

You should be able to use placeholders to dynamically change the value of configuration options when starting the server.

> For that, you should use the following format when defining the value of an option:

Reviewer: Try to follow "this" or "that" with a noun ....   For that purpose, 
Also, generally convert "You should use" to "Use..."  It means the same thing and is more concise.

```
db.username=${r"${my.db.user}"}
```

> By using a placeholder like in the example above, the server is going to try to resolve the `my.db.user` variable from:

Reviewer: Generally replace "like" with "such as"

[NOTE]
You should not use placeholders for build options. The reason is that these options
can not change once they are persisted into the server image. Prefer using placeholders only for the configurations options available to
the `start` command.

=== Understanding the default profiles

By default, the server defines two main profiles:

* `dev`
* `prod`

The `dev` profile is activated every time you run the `start-dev` command. In this profile, some key configuration options are set to make it possible to start the
server for development purposes without the burden of having to define additional settings that are mandatory for production.

The `prod` profile is activated by default when you run the `build` or the `start` command. You should use this profile to set any configuration option that
makes sense when deploying the server in production.

By default, the configurations options for the `prod` profile are commented in the `conf/keycloak.conf`. By looking at these options
you should have an idea about the main settings that you should consider when running in production.

> You should be able to change the configuration options for these built-in profiles as well as create your own.

Reviewer: "should be able to" sounds uncertain.   How about "can"

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/kc.adoc" as kc>

> <#import "/templates/options.adoc" as opts>

Reviewer 2: can be removed, unused

<#import "/templates/links.adoc" as links>

<@tmpl.guide
title="Configuring Keycloak"
summary="An overview of the server configuration">

In this guide, you will learn the core concepts around the server configuration and how to configure the server using the different
server options available. You will also learn how to properly configure the server to achieve an optimal runtime for faster
startup and low memory footprint.

== Understanding Server Options

> All the server options can be found by using the `help` option for individual commands:

Reviewer 2: ```suggestion
All server options can be found by using the `help` option for individual commands:
```

Author: I changed that based on Andrew comment already.

Reviewer 2: not hard on this, and works for me when @andymunro says it's better :)

> <@kc.all parameters="--help"/>

Reviewer 2: having read the below, i'd add this whole part around line 58, where you mention the 2 stages first and also command-specific options.

> <@kc.all parameters="--help"/>

Reviewer 3: I agree, I tend to not like the @kc.all as it creates a messy/confusing example

> You can instead look at all the server options at <@links.server id="all-config"/>.

Reviewer 2: ```suggestion
Alternatively, the <@links.server id="all-config"/> guide shows all available server options.
```

> Server options are loaded from different sources in a specific order and they use different formats:

Reviewer 2: I'd add a sentence about the ordinal concept, which makes clear that CLI takes precedence over Environment takes precedence over config file. I know that the table should show this, but I think it does not get clear enough for people looking at it the first time without prior knowledge.

Reviewer 3: I agree, it's not quite explicit if it's first in the list, or last in the list that takes precendence. I would assume it's first, but could also misinterpret that

You can also check all the options that require a build by looking at the `help` message of the `build` command:

<@kc.build parameters="--help"/>

The `build` command is responsible for producing an immutable and optimized server image, which is similar to building a container image. In addition to persisting
any build option you have set, this command also performs a series of optimizations to deliver the best runtime when starting and running the server. As a result,
a lot of processing that would usually happen when starting and running the server is no longer necessary and the server can start and run faster.

Some optimizations performed by the `build` command are:

* Closed-world assumption about installed providers, which means that no need to re-create the registry every time the server starts
* Configuration files are pre-parsed to reduce IO when starting the server
* Database specific resources are configured and prepared to run against a specific database vendor
* By persisting build options into the server image, the server does not perform any additional step to interpret configuration options and (re)configure itself

> Once you run the `build` command, you won't need to set the same build options again when starting the server.

Reviewer 3: ```suggestion
Once you run the `build` command, you won't need to set the same **build options** again when starting the server.
```

.First stage, run the `build` command to set any build option
<@kc.build parameters="--db=postgres"/>

The second stage involves starting the server using any **configuration option** available from the `start` command.

.Second stage, run the `start` command to set any configuration option when starting the server
<@kc.start parameters="--db-url-host=keycloak-postgres --db-username=keycloak --db-password=change_me --hostname mykeycloak.acme.com"/>

At this stage, you are free to set any value you want to any of the configuration options.

== Configuring the server for an optimal startup time

In addition to the optimizations performed when you run the `build` command, you might want to avoid using CLI options when running the
`start` command in favor of using environment variables or configuration options from the `conf/keycloak.conf` file.

.Set any build option
<@kc.build parameters="--db=postgres"/>

.Set any configuration option to `conf/keycloak.conf`
```
db-url-host=keycloak-postgres
db-username=keycloak
db-password=change_me
hostname mykeycloak.acme.com
```

.Start the server
<@kc.start/>

> By following that approach, the server is going to skip some steps at startup and start faster.

Reviewer 3: This section doesn't make any sense to me as it is suggesting properties in `conf/keycloak.conf` are somehow optimised compared to environment variables and configuration options, which is not the case. You've already previously mentioned that people can mix/match what config source they want, as well as the build stuff, so I would just remove this section completely as it's just duplication.

Author: @stianst It is the case. By not using CLI options the server skips some additional steps related to CLI init/parsing.

I'm only using the configuration file as an example, but the section also mentions using env vars.

I don't think we should remove this section.

Reviewer 3: Got a bit confused when reading the source, when I read it in html file format it made more sense

== Using the `auto-build` option when starting the server

The `auto-build` option allows you to perform the two configuration stages using a single command. Under certain circumstances, you might want to sacrifice the server startup time and update the values of build options when starting the server.

For that, you can start the server as follows:

.Using the `auto-build` option
<@kc.start parameters="--auto-build --db postgres --db-url-host keycloak-postgres --db-username keycloak --db-password change_me --hostname mykeycloak.acme.com"/>

> By using this option, the server is going to calculate the build options that have changed and automatically run the `build` command, if necessary, before actually starting the server.

Reviewer 2: ```suggestion
By using this option, the server is going to calculate the build options that have changed and automatically runs the `build` command, if necessary, before starting the server.
```

== Configuring the server using configuration files

By default, the server is going to always fetch any configuration option you set from the `conf/keycloak.conf` file. When you are using a fresh distribution,
this file holds only the recommended settings for running in production, which are initially commented out.

You can also specify a different configuration file by using the `[-cf|--config-file] option as follows:

<@kc.all rootParameters="-cf myconfig.conf"/>

> Changes to any *build option* defined in the `keycloak.conf` file that is targeted for the `build` command is ignored

Reviewer 2: ```suggestion
Changes to any *build option* defined in the `keycloak.conf` file that is targeted for the `build` command are ignored
```

if the value differs from the value used to previously run the `build` command. In this case, make sure you run the `build` command again so that
any build option is updated accordingly.

== Using placeholders

You are able to use placeholders to dynamically change the value of configuration options when starting the server.

For that purpose, you use the following format when defining the value of an option:

```

> db.username=${r"${my.db.user}"}

Reviewer 3: I don't understand this syntax/example.

What's the $r stuff? Can't I just do db-username=${my.db.user}?

Also, it's db-username, not db.username

Reviewer 2: Fell into that, too. It's telling the parser to parse this **r*aw, so look at the rendered output, you'll see it's ${my.db.user}

Reviewer 3: Got no clue what that means, and reckon pretty much everyone reading this will be confused as well

Author: This is how we can escape special characters in freemarker.

```

By using a placeholder such as in the example above, the server is going to try to resolve the `my.db.user` variable from:

* A system property using the same key

> * An environment variable by replacing dots with `_`

Reviewer 3: Should have examples for both, but then again we don't allow you to easily set system properties, so I don't know how I would use that now. So, perhaps we should just stick to documenting only environment variables?

Author: Makes sense.

> You should not use placeholders for build options. The reason is that these options

Reviewer 3: If you should not use placeholders for build options, why do we support it?

Author: Changed that part to remove this text and make sure only env vars are documented.

can not change once they are persisted into the server image. Prefer using placeholders only for the configurations options available to
the `start` command.

> === Understanding the default profiles

Reviewer 3: I'd change this section to cover just the dev mode, rather than different "profiles".

Author: Not sure, I think we should also cover the default *configuration* profiles we have. Both dev and prod.

> === Understanding the default profiles

Reviewer 2: ```suggestion
=== Understanding development mode and production mode
```

Reviewer 2: that would mean there are / may be multiple production modes / development modes as far as i can tell. @andymunro I am not sure aboout this, though. Want to avoid ambiguity, there are exactly 2 modes, development and production. wdyt, as a native speaker? ;)

1) "Understanding development mode and production mode"
2) "Understanding development and production mode"
3) "Understanding development and production modes"
4) "Understanding development- and production mode"
5) "Understanding development- and production modes" 

These are the options floating through my head, but not sure what is exactly right english tbh

> By default, the server defines two main profiles:

Reviewer 2: ```suggestion
By default, the server defines two main operating modes:
```

> * `dev`

Reviewer 2: Development 
```suggestion
* Development mode, using the `start-dev` command
```

> * `prod`

Reviewer 2: ```suggestion
* Production mode mode, using the `start` command
```

> The `dev` profile is activated every time you run the `start-dev` command. In this profile, some key configuration options are set to make it possible to start the

Reviewer 2: ```suggestion
The development mode is activated every time you run the `start-dev` command. In this mode, some key configuration options are set to make it possible to start the
```

server for development purposes without the burden of having to define additional settings that are mandatory for production.

> The `prod` profile is activated by default when you run the `build` or the `start` command. You should use this profile to set any configuration option that

Reviewer 2: ```suggestion
The production mode is activated by default when you run the `build` or the `start` command. Use this mode to set any configuration option that
```

> makes sense when deploying the server in production.

Reviewer 2: ```suggestion
is needed for deploying Keycloak in production.
```

> By default, the configurations options for the `prod` profile are commented in the `conf/keycloak.conf`. By looking at these options

Reviewer 2: ```suggestion
By default, the configurations options for the production mode are commented out in the `conf/keycloak.conf`. These examples
```

> you should have an idea about the main settings that you should consider when running in production.

Reviewer 2: ```suggestion
 are meant to give you an idea about the main settings that needs to be considered when running in production.
```

configuration options available. You will also learn how to properly configure the server to achieve an optimal runtime for faster
startup and low memory footprint.

== Configuration Options

All the available configuration options can be found by either running the `help` command for individual commands:

<@kc.all parameters="--help"/>

Or by looking at all the configuration options at the <@link.guide title="All configuration"/> section.

These options can be set using different formats:

.Command-line argument
<@kc.all parameters="--<category>-<sub-category>-<property>=<value>"/>

.Environment variable
```
export KC_CATEGORY_SUB_CATEGORY_PROPERTY=<value>
```

.Properties at `conf/keycloak.properties`
```
<category>.<sub-category>.<property>=<value>
```

The format you should use is use-case specific but most of the time it should be enough to use the CLI. As an example, when running
the server as a container you should be fine using both environment variables or command-line options.

[NOTE]
If you are unsure about the format of a specific property, consider looking at the <@link.guide title="All configuration"/> section.

> If a same configuration option is defined in different configuration sources, the effective value should be resolved respecting the following order:

Author: I'm definitely not a native speaker. I agree about `the`/`will` but is `by the following order` correct? Or do you mean `by respecting the following order`?

> If a same configuration option is defined in different configuration sources, the effective value should be resolved respecting the following order:

Reviewer 2: no native speaker here, too, so tbh I don't know for sure. Let's leave the 'by' it as is and only change the others

. Command-line arguments
. Environment variables
. Properties

== Configuring the server for optimal runtime

Some configuration options are narrowed to a specific command or configuration stage. This clear separation is one of the key aspects
of the server configuration and related to a series of optimizations that are done in order to deliver the best runtime when starting and running the server.

By looking at the individual configuration options at this <@link.guide title="All configuration" text="section"/>, you will notice
that some options require a `build`. In other words, they can only be set when running the `build` command.
You can also check all the options that require a build by looking at the `help` message of the `build` command:

<@kc.build parameters="--help"/>

[NOTE]

> By default, the `help` message for the `build` command only shows the options that require a build. You can also use the `help-all` option to list all the available options, including those

Author: I like the way it is and aligned with our previous discussions. One of the main points of having both `help` and `help-all` is that we have a clear separation between what is targeted and specific for a specific command/step.

Where `help-all` give you everything.

> By default, the `help` message for the `build` command only shows the options that require a build. You can also use the `help-all` option to list all the available options, including those

Reviewer 2: ok np

that do not require a build.

The `build` command is responsible for producing an immutable and optimized server image (similar concept to building a container image). In addition to persisting
any configuration option you have set, this command will also perform a series of optimizations to deliver the best runtime when starting and running the server. Basically,

> a lot of processing that would usually happen when starting/running the server is no longer necessary and the server can start and run faster.

Reviewer 2: "is no longer necessary at startup" would be a better explanation imo. It is still necessary, just happens beforehand at the build step.

Author: Yeah, but it does not happen when "starting/running" the server. That is my point.

What do you suggest?

Reviewer 2: perhaps 
"Basically, a lot of processing that would usually happen when starting/running the server is now happening beforehand using `build`, so the server can start and run faster." ?

Reviewer 2: +1, works for me.

<@kc.start/>

As a rule of thumb, prefer building a server image to achieve the optimal runtime as well as running the `start` command without any parameter for faster startup time.

=== Overriding settings

Once you have built a server image using the `build` command, you should be able to override any property that is available from the `start` command.

For instance, if you have built an image as follows:

<@kc.build parameters="--db=postgres --db-url-host=keycloak-postgres --db-username=keycloak --db-password=change_me --hostname mykeycloak.acme.com"/>

You should be able to override the `db-username`, `hostname`, or any other property available to the `start` command:

<@kc.start parameters="--db-username foo --hostname newkeycloak.acme.com"/>

As a result, the server should start using all the configuration options previously set using the `build` command but overriding those that you explicitly
set when starting the server.

This configuration approach can be useful when building an opinionated server image with the possibility to override specific options that are allowed to
change when starting the server (those available to the `start` command).

== Using placeholders

> You should be able to use placeholders to dynamically change the value of configuration options when starting the server.

Reviewer 2: You 'are' able? Because people are, actually. So it's shorter, and more concise ;)

For that, you should use the following format when defining the value of an option:

```

> db.username=${r"${my.db.user}"}

Reviewer 2: When I use this syntax, I get 
```
ERROR: Build failure: Build failed due to errors
	[error]: Build step org.keycloak.quarkus.deployment.KeycloakProcessor#persistBuildTimeProperties threw an exception: java.util.NoSuchElementException: SRCFG00011: Could not expand value r"mydbuser" in property kc.db.username
```

Typo? Should be ${my.db.user}, shouldn't it? Also: I'd prefer the example to be with an environment variable instead of a system property. So ${MY_DB_USER}. Imo we shouldn't encourage the usage of system properties, but that's just my opinion.

Author: If you try to render the doc you should get what you want. Could you please check?

W.r.t. using env vs props, in general I agree. But here we are explaining that placeholders can be resolved from both sys props and env vars.

Otherwise we would be limiting to only env vars?

```

By using a placeholder like in the example above, the server is going to try to resolve the `my.db.user` variable from:

* A system property using the same key
* An environment variable by replacing dots with `_`.

Placeholders can also be used when building a server image using the `build` command so that the value of a configuration option is updated
with the value set to the placeholder (system property or environment variable) when starting the server.

[NOTE]
You should not use placeholders for configuration options that can only be set through the `build` command. The reason is that these options
can not change once they are persisted into the server image. Prefer using placeholders only for the configurations options available to
the `start` command.

== Configuration profiles

Most of the time the server is deployed to different environments like testing, staging, and production environments. Although they share some
core configuration options (e.g: database, clustering, etc) some other options might change on a per-environment basis.

> In order to make easier to support multiple configurations, the server allows you to create configuration profiles.

Reviewer 2: In order to make _it_ easier to..., _Keycloak_ allows... ?

Author: If I have to change `server` then I would need to change a lot of other places. The reason why I'm not using `Keycloak` is to be more neutral about community/product. We should probably have a macro to resolve the proper name similarly as how we do in the current documentation?

Reviewer 2: works for me :) but that macro is a good idea, in general.

Configuration profiles are basically a set of one or more configuration options grouped together using a unique key.

You can activate a profile by using the `[-pf|--profile]` option as follows:

<@kc.all rootParameters="-pf myprofile"/>

Once a profile is activated, all the configuration options defined for the profile should take effect when configuring or starting the server.

=== Understanding the default profiles

By default, the server defines two main profiles:

> * `dev`

Reviewer 2: shall we mention the 'test' profile which should not be used bc. it's internal?

Author: I don't think so. Just because we don't have a default `test` profile.

But that is a good point. TBH, I do not know the impact of activating the `test` profile due to the changes we did to our test suite. I mean, people should be able to use `test` as a key.

Reviewer 2: Yes, in general it works, see attached screenshot, and I'm also not sure if we should discourage ppl to use test as key.
![Screenshot 2022-01-11 at 20 13 18](https://user-images.githubusercontent.com/89905860/149006630-94d53182-48d5-48a6-b5b4-9c99bd248f37.png)

Author: We should not.

* `prod`

The `dev` profile is activated everytime you run the `start-dev` command. In this profile, some key configuration options are set to make possible to start the
server for development purposes without the burden of having to define additional settings that are mandatory for production.

The `prod` profile is activated by default whenever you run the `build` or the `start` command. You should use this profile to set any configuration option that
makes sense when deploying the server in production.

By default, the configurations options for the `prod` profile are commented in the `conf/keycloak.properties`. By looking at these options
you should have an idea about the main settings that you should consider when running in production.

You should be able to change the configuration options for these built-in profiles as well as create your own.

=== Defining a configuration profile

A profile can be created by adding a set of one or more properties to the `conf/keycloak.properties` file as follows:

```
%myprofile.db.username=foo
%myprofile.db.password=foo_secret
%myprofile.hostname=mykeycloak
```

> Any configuration option associated with a profile should be prefixed by a `%` followed by the profile name and a dot `.` like in the example above.

Reviewer 2: "If no profile is given for a specific property, the global property is used instead, so take care what you add to the global namespace" or sth?

Author: Good point. Going to add a note.

== Using Quarkus configuration options

The server is built on top of <@link.external url="https://quarkus.io" text="Quarkus"/> and you should be able to use any property available from Quarkus

> if you do not find a corresponding options in the server configuration. The server has a configuration layer that provides a 1:N mapping between the server configuration options and those available from Quarkus.

Reviewer 2: Imo this is too detailed for a user guide. Imo the important thing here is: "I want to know how I can use quarkus properties", not "what is our configuration layer doing internally".

Reviewer 2: So a short example, together with the statement that we'd prefer to have a discussion on why ppl need it and add it to our layer would be better i think.

Author: Yeah, kind of exposing a bit of details. But the idea is to express that we do support Quarkus properties and if you are doing that is because we can probably improve our configuration.

I do not see an example helping much.

Or perhaps we should remove this entirely and move to some other guide?

Reviewer 2: +1 for removing completely, for now. Only generates noise imo, as is. :)

Alternatively, I thought about sth like this:

```
== Using Quarkus configuration options
The server is built on top of <@link.external url="https://quarkus.io" text="Quarkus"/> and you are basically able to use any property available within Quarkus, using for example `quarkus.datasource.jdbc.max-size=...`. 

We generally want to provide an extra convenience layer over the raw Quarkus properties using our own configuration. So we'd love to hear from you over at <@link.external url="https://github.com/keycloak/keycloak/discussions" text="GitHub Discussions"/>. if you think something is missing.
``` 

Either way fine for me ;)

== Configuring the server using configuration files

By default, the server is going to always fetch any configuration option you set from the `conf/keycloak.conf` file. When you are using a fresh distribution,
this file holds only the recommended settings for running in production, which are initially commented out.

You can also specify a different configuration file by using the `[-cf|--config-file] option as follows:

.Running the `build` command using a custom configuration file
<@kc.build rootParameters="-cf myconfig.conf"/>

.Running the `start` command using a custom configuration file
<@kc.start rootParameters="-cf myconfig.conf"/>

Changes to any *build option* defined in the `keycloak.conf` file that is targeted for the `build` command are ignored
if the value differs from the value used to previously run the `build` command. In this case, make sure you run the `build` command again so that
any build option is updated accordingly.

== Using placeholders

You are able to use placeholders to dynamically change the value of configuration options when starting the server.

For that purpose, you use the following format when defining the value of an option:

```

> db.username=${r"${my.db.user}"}

Reviewer 3: ```suggestion
db-username=${r"${my.db.user}"}
```

> db.username=${r"${my.db.user}"}

Reviewer 3: I would have expected it to be something like:
```
db-username=${MY_ENV_VALUE}
```

Author: Yeah, accepted your suggestion.

> == Using placeholders
> 
> You are able to use placeholders to dynamically change the value of configuration options when starting the server.
> 
> For that purpose, you use the following format when defining the value of an option:
> 
> ```
> db-username=${r"${my.db.user}"}
> ```
> 
> By using a placeholder such as in the example above, the server is going to try to resolve the `my.db.user` placeholder from the `KC_MY_DB_USER` environment variable.
> 

Reviewer 3: ```suggestion
```

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/kc.adoc" as kc>
<#import "/templates/options.adoc" as opts>
<#import "/templates/link.adoc" as link>

<@tmpl.guide

> title="Configuring Keycloak"

Reviewer: "Configuring {project_name}"

summary="An overview of the server configuration">

In this guide, you will learn the core concepts around the server configuration and how to configure the server using the different
configuration options available. You will also learn how to properly configure the server to achieve an optimal runtime for faster
startup and low memory footprint.

== Configuration Options

All the available configuration options can be found by either running the `help` command for individual commands:

<@kc.all parameters="--help"/>

Or by looking at all the configuration options at the <@link.guide title="All configuration"/> section.

These options can be set using different formats:

.Command-line argument

> <@kc.all parameters="--<category>-<sub-category>-<property>=<value>"/>

Reviewer 3: I'd remove use of @kc.all as it's confusing

> <@kc.all parameters="--<category>-<sub-category>-<property>=<value>"/>

Author: I've added more details about the format. See if helps. 

IMO, it is important to give some info about how an option should/is named.

> <@kc.all parameters="--<category>-<sub-category>-<property>=<value>"/>

Author: Removed

.Environment variable
```

> export KC_CATEGORY_SUB_CATEGORY_PROPERTY=<value>

Author: Removed

```

.Properties at `conf/keycloak.properties`
```
<category>.<sub-category>.<property>=<value>
```

The format you should use is use-case specific but most of the time it should be enough to use the CLI. As an example, when running
the server as a container you should be fine using both environment variables or command-line options.

[NOTE]
If you are unsure about the format of a specific property, consider looking at the <@link.guide title="All configuration"/> section.

> If the same configuration option is defined in different configuration sources, the effective value will be resolved by respecting the following order:

Reviewer: change "will be" to "is:
in most cases, present tense works. It sounds more active, but means the same thing.

. Command-line arguments
. Environment variables
. From `conf/keycloak.properties`

== Configuring the server for optimal runtime

Some configuration options are narrowed to a specific command or configuration stage. This clear separation is one of the key aspects

> of the server configuration and related to a series of optimizations that are done in order to deliver the best runtime when starting and running the server.

Reviewer: add  "is" before "related"  which reads better

> By looking at the individual configuration options at this <@link.guide title="All configuration" text="section"/>, you will notice

Reviewer: remove "will"

that some options require a `build`. In other words, they can only be set when running the `build` command.
You can also check all the options that require a build by looking at the `help` message of the `build` command:

<@kc.build parameters="--help"/>

[NOTE]
By default, the `help` message for the `build` command only shows the options that require a build. You can also use the `help-all` option to list all the available options, including those
that do not require a build.

The `build` command is responsible for producing an immutable and optimized server image (similar concept to building a container image). In addition to persisting

> any configuration option you have set, this command will also perform a series of optimizations to deliver the best runtime when starting and running the server. Basically,

Reviewer: "command also performs..."

> a lot of processing that would usually happen when starting/running the server is no longer necessary and the server can start and run faster.

Reviewer: starting or running

Some optimizations performed by the `build` command are:

> * Closed-world assumption about installed providers, hence no need to re-create the registry everytime the server starts

Reviewer: every time

> * Configuration files (e.g.: clustering) are pre-parsed to reduce IO when starting the server

Reviewer: Generally avoid latin: e.g.
This should either be "for example" or "in other words"

* Database specific resources are configured and prepared to run against a specific database vendor
* By persisting configuration options into the server image, the server does not perform any additional step to interpret configuration options and (re)configure itself

> There are many more optimizations, most of them performed by <@link.external url="https://quarkus.io" text="Quarkus"/> itself. Just keep in mind that this additional step is a key aspect to achieve the optimal runtime,

Reviewer: Avoid "There are" which makes no sense grammatically.  What is the subject of the sentence.  In general use a phrase about something that exists, such as "Many more optimizations exist and most of them..."

You should be able to use placeholders to dynamically change the value of configuration options when starting the server.

For that, you should use the following format when defining the value of an option:

```
db.username=${r"${my.db.user}"}
```

By using a placeholder like in the example above, the server is going to try to resolve the `my.db.user` variable from:

* A system property using the same key
* An environment variable by replacing dots with `_`.

Placeholders can also be used when building a server image using the `build` command so that the value of a configuration option is updated
with the value set to the placeholder (system property or environment variable) when starting the server.

[NOTE]
You should not use placeholders for configuration options that can only be set through the `build` command. The reason is that these options
can not change once they are persisted into the server image. Prefer using placeholders only for the configurations options available to
the `start` command.

== Configuration profiles

Most of the time the server is deployed to different environments like testing, staging, and production environments. Although they share some

> core configuration options (e.g: database, clustering, etc) some other options might change on a per-environment basis.

Reviewer: change (e.g: database, clustering, etc)
TO
, such as database or clustering,

In order to make it easier to support multiple configurations, the server allows you to create configuration profiles.

Configuration profiles are basically a set of one or more configuration options grouped together using a unique key.

You can activate a profile by using the `[-pf|--profile]` option as follows:

<@kc.all rootParameters="-pf myprofile"/>

Once a profile is activated, all the configuration options defined for the profile should take effect when configuring or starting the server.

[NOTE]
If no profile is given for a specific property, the configuration option is applied to all profiles.

=== Understanding the default profiles

By default, the server defines two main profiles:

* `dev`
* `prod`

> The `dev` profile is activated everytime you run the `start-dev` command. In this profile, some key configuration options are set to make possible to start the

Reviewer: every time

server for development purposes without the burden of having to define additional settings that are mandatory for production.

> The `prod` profile is activated by default whenever you run the `build` or the `start` command. You should use this profile to set any configuration option that

Reviewer: Change "whenever" to "when"  It means the same and it is shorter.

makes sense when deploying the server in production.

By default, the configurations options for the `prod` profile are commented in the `conf/keycloak.properties`. By looking at these options
you should have an idea about the main settings that you should consider when running in production.

You should be able to change the configuration options for these built-in profiles as well as create your own.

=== Defining a configuration profile

A profile can be created by adding a set of one or more properties to the `conf/keycloak.properties` file as follows:

```
%myprofile.db.username=foo
%myprofile.db.password=foo_secret
%myprofile.hostname=mykeycloak
```

> Any configuration option associated with a profile should be prefixed by a `%` followed by the profile name and a dot `.` like in the example above.

Reviewer: Change "like" to "such as"
